bluesky/callbacks/core.py

```diff
@@ -50,6 +78,125 @@ def stop(self, doc):
+    RunRouter will drop all its references to ``cb``. It is up to
+    ``callback_factory`` whether to return a new callable each time (having a
+    lifecycle of one run, garbage collected thereafter), or to return the same
+    object every time, or some other pattern.
+    
+    To summarize, the RunRouter's promise is that it will call each
+    ``callback_factory`` with each new RunStart document and that it will
+    forward all other documents from that run to whatever ``callback_factory``
+    returns (if not None).
+
+    Parameters
+    ----------
+    callback_factories : list
+        A list of callables with the signature:
+
+            callback_factory(start_doc)
+
+        which should return ``None`` or another callable with this signature:
+
+            callback(name, doc)
+    """
+    def __init__(self, callback_factories):
+        self.callback_factories = callback_factories
+        self.callbacks = defaultdict(list)  # start uid -> callbacks
+        self.descriptors = {}  # descriptor uid -> start uid
+        self.resources = {}  # resource uid -> start uid
+
+    def _event_or_bulk_event(self, doc):
+        descriptor_uid = doc['descriptor']
+        try:
+            start_uid = self.descriptors[descriptor_uid]
+        except KeyError:
+            # The belongs to a run that we are not interested in.
+            return []
+        return self.callbacks[start_uid]
+
+    def event(self, doc):
+        for cb in self._event_or_bulk_event(doc):
+            cb('event', doc)
+
+    def bulk_event(self, doc):
+        for cb in self._event_or_bulk_event(doc):
+            cb('bulk_event', doc)
+
+    def _datum_or_bulk_datum(self, doc):
+        resource_uid = doc['resource']
+        try:
+            start_uid = self.resources[resource_uid]
+        except KeyError:
+            # The belongs to a run that we are not interested in.
+            return []
+        return self.callbacks[start_uid]
+
+    def datum(self, doc):
+        for cb in self._datum_or_bulk_datum(doc):
+            cb('datum', doc)
+
+    def bulk_datum(self, doc):
+        for cb in self._datum_or_bulk_datum(doc):
+            cb('bulk_datum', doc)
+
+    def descriptor(self, doc):
+        start_uid = doc['run_start']
+        cbs = self.callbacks[start_uid]
+        if not cbs:
+            # This belongs to a run we are not interested in.
+            return
+        self.descriptors[doc['uid']] = start_uid
+        for cb in cbs:
+            cb('descriptor', doc)
+
+    def resource(self, doc):
+        start_uid = doc['run_start']
+        cbs = self.callbacks[start_uid]
+        if not cbs:
+            # This belongs to a run we are not interested in.
+            return
+        self.resources[doc['uid']] = start_uid
+        for cb in cbs:
+            cb('resource', doc)
+
+    def start(self, doc):
+        for callback_factory in self.callback_factories:
+            cb = callback_factory(doc)
```

Comment:
Do we need a mechanism for pushing more information into the `callback_factory`? One could imagine wanting to pass information into the callback at init time which is not in the document. For instance, if I build a `callback_factory` which returns a `LiveScatter` how do I change the colormap of the scatter plot for the next scan?

Reply:
Maybe the `callback_factory` signatures could be `cf(start_doc, **kwargs)` and we pass on **kwargs from the `RunRouter` itself? This would require the factory to take in arbitrary kwargs and pick and choose among them but that might be ok.